cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/ExecutionSparkContext.java

```diff
@@ -113,18 +118,95 @@ public ExecutionSparkContext(SparkSpecification specification, Id.Program progra
 
   @Override
   public <T> T readFromDataset(String datasetName, Class<?> kClass, Class<?> vClass) {
-    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
-
     // Clone the configuration since it's dataset specification and shouldn't affect the global hConf
     Configuration configuration = new Configuration(hConf);
+
+    // first try if it is InputFormatProvider
+    Map<String, String> dsArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+    Dataset dataset = getDataset(datasetName, dsArgs);
+    if (dataset instanceof InputFormatProvider) {
+      // get the input format and its configuration from the dataset
+      String inputFormatName = ((InputFormatProvider) dataset).getInputFormatClassName();
+
+      // load the input format class
+      if (inputFormatName == null) {
+        throw new DatasetInstantiationException(
+          String.format("Dataset '%s' provided null as the input format class name", datasetName));
+      }
+      try {
+        @SuppressWarnings("unchecked")
+        Class<? extends InputFormat> inputFormatClass =
+          (Class<? extends InputFormat>) SparkClassLoader.findFromContext().loadClass(inputFormatName);
+        Map<String, String> inputConfig = ((InputFormatProvider) dataset).getInputFormatConfiguration();
+        if (inputConfig != null) {
+          for (Map.Entry<String, String> entry : inputConfig.entrySet()) {
+            configuration.set(entry.getKey(), entry.getValue());
+          }
+        }
+        return getSparkFacade().createRDD(inputFormatClass, kClass, vClass, configuration);
+
+      } catch (ClassNotFoundException e) {
+        throw new DatasetInstantiationException(String.format(
+          "Cannot load input format class %s provided by dataset '%s'", inputFormatName, datasetName), e);
+      } catch (ClassCastException e) {
+        throw new DatasetInstantiationException(String.format(
+          "Input format class %s provided by dataset '%s' is not an input format", inputFormatName, datasetName), e);
+      }
+    }
+
+    // it must be supported by SparkDatasetInputFormat
+    Map<String, String> datasetArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
     SparkDatasetInputFormat.setDataset(configuration, datasetName, datasetArgs);
     return getSparkFacade().createRDD(SparkDatasetInputFormat.class, kClass, vClass, configuration);
   }
 
   @Override
   public <T> void writeToDataset(T rdd, String datasetName, Class<?> kClass, Class<?> vClass) {
+    // Clone the configuration since it's dataset specification and shouldn't affect the global hConf
+    Configuration configuration = new Configuration(hConf);
+
+    // first try if it is OutputFormatProvider
+    Map<String, String> dsArgs = RuntimeArguments.extractScope(Scope.DATASET, datasetName, getRuntimeArguments());
+    Dataset dataset = getDataset(datasetName, dsArgs);
```

Comment:
And to just get a new dataset instance without putting into the `datasets` instance cache, you can call `instantiateDataset` instead of `getDataset` (and that was why the `instantiateDataset` exists, which you can see both the `getBatchReadable` and `getBatchWritable` calls instead of `getDataset`).